ignitionrobotics/ign-gazebo
# Set a cloned joint's parent/child link names to the cloned parent/child link names

Signed-off-by: Ashton Larkin <ashton@openrobotics.org>


# 🦟 Bug fix

## Summary
When a joint is cloned, the cloned joint's parent and child links need to be set to the cloned parent and child links (currently, the cloned joint's parent/child links are the original joint's parent/child links). Otherwise, links will not be connected to joints when cloning things like models that have links and joints.

The approach taken here is similar to the approach taken for updating a cloned canonical link in #959, where internal maps are used to keep track of a joint's original links and cloned links. As discussed in https://github.com/ignitionrobotics/ign-gazebo/pull/959#discussion_r711385128, we shouldn't give special treatment to certain components. This is just a temporary workaround until we find a better solution.

To test this, do the following:
1. Start gazebo with a double pendulum, since this model has links and joints:
```
ign gazebo -r pendulum_links.sdf
```
2. In another terminal window, clone the double pendulum:
```
ign service -s /world/double_pendulum/create --reqtype ignition.msgs.EntityFactory --reptype ignition.msgs.Boolean --timeout 300 --req 'clone_name: "double_pendulum_with_base", pose: {position: {y: 3}}, name: "pendulum_copy"'
```
3. You should now see two pendulums operating as expected:

![functioning_joints](https://user-images.githubusercontent.com/42042756/135555463-ab6e4267-cfae-4bff-9a0b-91a75a582295.gif)

If you repeat the steps above without this PR, you'll notice that the links on the cloned pendulum fall to the ground because the cloned pendulum does not have proper joints:

![bad_joints_clone](https://user-images.githubusercontent.com/42042756/135555025-ea9e5999-e7c5-44ee-9988-f203b0abf907.gif)


## Checklist

**Note to maintainers**: Remember to use **Squash-Merge**
